fastlane/fastlane
# [Ruby 3.0] replace slack-notifier gem with internal library

### Checklist

### Motivation and Context



As per the plan #17931 for Ruby 3.0, I've been working on fixing Ruby 2.7's deprecation warnings which will crash on Ruby 3.0. There is an issue where one of the fastlane's dependencies slack-notifier, which is not maintained any more, keeps producing that warning. As I explained in https://github.com/fastlane/fastlane/pull/18512 I decided to go with having our own implementation to post messages to Slack in fastlane. That way we get full control of logic. It is not only about Ruby version support but also Slack's webhook transition.

Right now there are two translations ongoing 

* Legacy incoming webhook https://api.slack.com/legacy/custom-integrations/messaging/webhooks vs The latest incoming webhook using Slack app https://api.slack.com/messaging/webhooks 
* Legacy attachments in formatting https://api.slack.com/reference/messaging/attachments vs BlockKit https://api.slack.com/messaging/composing/layouts#sending-messages

To be fair, `slack-notifier` does support those but we don't know what changes will be coming from Slack in future.

### Description



Prior to this PR, I've updated `fastlane/spec/fastlane/actions_specs/slack_spec.rb` so it's now independent from the gem. This PR added `Fastlane::Notification::Slack` class to own simple alternative of `slack-notifier`. (I suppose it might be interesting to support Microsoft teams or whatever other popular chat tools in fastlane as it's remote working era🙂 That's why it's got `Fastlane::Notification` namespace.) 

After digging through the gem, I realised most of [the extra features](https://github.com/stevenosloan/slack-notifier/tree/master/lib/slack-notifier/payload_middleware) in `slack-notifier` was not used by fastlane. Only `Slack::Notifier::Util::LinkFormatter` was needed to be re-implemented.

### Testing Steps



Get an incoming webook url from here
https://www.slack.com/apps/manage/custom-integrations

Add this snippet to `fastlane/Fastfile` in the repo

```ruby
lane :slack_test do
  slack(
    slack_url: 'https://hooks.slack.com/services/xxxxxxxxxxx',
    pretext: 'pretext test <@xxxxxx> #general',
    username: 'ainame-bot',
    channel: '#general',
    message: 'test <@xxxxxx> #general  <a href="https://fastlane.tools">fastlane</a> [GitHub](https://github.com/fastlane)',
    link_names: true,
  )
end
```

Run `bundle exec fastlane slack_test`

Legacy incoming webhook | Slack App's incoming webhook
--|--
<img width="690" alt="Screenshot 2021-04-10 at 08 20 29" src="https://user-images.githubusercontent.com/748949/114262035-9f063500-99d5-11eb-8186-16c73956c52d.png"> | <img width="565" alt="Screenshot 2021-04-10 at 08 30 08" src="https://user-images.githubusercontent.com/748949/114262276-0375c400-99d7-11eb-8870-a2d92166d91c.png">

As you can see Slack app's incoming webhook can't overwrite username and icon. Instead, you need to customise them on its settings page.
